### PR TITLE
avoid adding duplicate/colliding param IDs

### PR DESCRIPTION
**Pull request by dndrks**
from https://llllllll.co/t/bounds/23336/68, the params system doesn't currently protect against ID collision, which can lead to unintentional overwrites. this implementation simply prints the error to matron, same as the other `paramset.add()` errors -- it seemed best to stick with this pattern, unless we want to actually force a load error?

i tested this approach with `bounds`, which positively surfaced the error and avoided the collision by not adding the script's `input_level` re-definition. also tested with scripts that have a lot of params and didn't see any noticeable load time lag with this iterative approach, but am totally interested/open to optimization of this rather brute-force approach!

### CODE REVIEW

**Review by tehn — APPROVED** [review]
looks good. i wouldn't force a load error--- your solution is sufficient and helpful for debugging

**Comment by catfact** [comment]
>  totally interested/open to optimization 

maybe you can use `lookup_param` instead of iterating. (we maintain this reverse LUT for exactly that reason. though of course it is not so critical here, as opposed to accessing the parameter at runtime.)

File: lua/core/paramset.lua [hunk 1]
```diff
@@ -118,6 +118,13 @@ function ParamSet:add(args)
     end
   end
 
+  for _,p in pairs(self.params) do
```
[the remark this answers is not in the document]
Comment (tehn) [reply lua/core/paramset.lua:121]: oof i forgot we have a reverse lookup. yes this is of course much more optimal

**Comment by dndrks** [comment]
actually, intercepting here with either approach doesn't stop overwriting the action -- i'll keep digging, apologies for the premature PR!

**Comment by catfact** [comment]
not sure i follow... you mean, someone can still do

```
-- add a param with existing ID
params:add("input_level", ...)
-- this returns nil, but the caller doesn't know / care

--- now this will overwrite the system parameter action
params:set_action("input_lvel", ...)
```

... but that's not the error we were seeing, is it? i agree it's an issue but not sure what we can do about it except add a mechanism to make parameter objects write-protected when they are added by the system.

the change still seems good, because regardless of what else the script does with overwriting actions, we _definitely_ don't want to continue past the early return that you added if the `id` is already a valid key in `self.lookup`. doing so means that the param count is wrong, `lookup` is wrong, etc.

**Comment by dndrks** [comment]
yep, that's it -- placing this return is just an alert, but it allows the script to still overwrite the action, which i guess i just mentally categorized not fixing the issue. but you're right that it's not an issue that likely would've happened if the author was alerted ahead of release.

> not sure what we can do about it except add a mechanism to make parameter objects write-protected when they are added by the system.

as someone who often adds lil' custom callbacks to system actions like clock source changes, etc in my scripts, def want to avoid that level of clamping down. perhaps i can add more info to the error, so script authors know that they're overwriting the previous action -- this should avoid trouble moving forward!

**Comment by catfact** [comment]
do you restore system-defined actions at cleanup?

(we could explicitly support the use case of adding to system parameter hooks; bonuhs = we handle cleanup)

**Comment by catfact** [comment]
i guess i'm not sure why we wouldn't want to throw an init error. 

i mean... at best, the script will probably not work quite right. (e.g., new collding parameter attributes will be meaningless, because attempts to access the parameter will get the existing one instead of the new one.) 

at worst, could mess up the system unrecoverably in some way. (if the system param gets removed at script cleanup, or its action gets overwritten.)

**Comment by tehn** [comment]
yeah, init error is fine, i was under-appreciating the issue.

**Comment by dndrks** [comment]
alright, rad, re-opening! thanks both for the insights + guidance!

> do you restore system-defined actions at cleanup?

this was a super helpful q, because it helped confirm that the system automatically cleans out system-defined params and re-adds them (+ their actions) with each script load: https://github.com/monome/norns/blob/4ea23d8f3ff2ca7c082c0e6f0302755d5bfd83c8/lua/core/script.lua#L94-L100. this also totally clarified some weird behavior i was seeing, where the lookup table check would fail with _system_ params during the script clear stage -- which meant that the lookup table wasn't being cleared out before rebuilding those params. so i added `self.lookup = {}` to the `ParamSet:clear()` action and everything works!

so, tl;dr, two things changed:
- `ParamSet:clear()` now nukes the `.lookup` table, since it gets rebuilt as system params are added back in
- the `.lookup` check now provides a script init error and prints notice to matron

lemme know if anything seems off!

**Comment by catfact** [comment]
awesome, thanks for cleaning that up.

yeah I had totally forgot that system parameters are torn down and rebuilt! great

I think this error is great to have. It's not hard to accidentally dupe param id even within a script...
